File: 백준/Silver/1149. RGB거리/RGB거리.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

/**
 * 코드2. 탑다운 방식의 dp 사용
 * 기존 코드에서 구현했던 점화식을 그대로 차용(dp[n]=d0[n-1]+min(n)) 후, 각 색깔별로의 최소 비용 값 구함
 * 탑다운 방식에서의 종료 조건은, 인덱스가 0일 때로 설정하여, 재귀의 끝에서 해당 색깔의 0레벨에서의 비용을 뱉어내도록 구현
 * 해당 코드에서는, 분기처리가 중요했는데, 초기에는 분기처리를 어떻게 할까 싶어, 함수를 각 색깔별로 하나씩을 만들어야되나 싶었지만,
 * 결국 n번째와 n-1번째가 색이 달라야된다는 조건을 만족시키기위해서는 각 함수별로 만들어도 소용 없고, 분기 처리를 해야된다는 것을 꺠달음
 * 
 * +) 바텀업 방식에서는 각 레벨별로, 특정 색을 갖고 있는 별로 비용을 수동적으로 구해주지만, 탑다운 방식에 재귀 호출에 따라,
 * 초기 n-1레벨 스타팅 색만 지정해준다면, 어떠한 색으로 0레벨이 끝나는지는 몰라도, 최소 비용을 자동적으로 구해줄 수 있음
 */
public class Main {
    static final int numberOfColor = 3;
    static final int INF = -1;
    static final int RED = 0;
    static final int GREEN = 1;
    static final int BLUE = 2;

    static int getMinimumCase(int[][] dp, int[][] cost, int depthIndex, int colorIndex) {
        if (depthIndex == 0) {
            return cost[0][colorIndex];
        }
        if (dp[depthIndex][colorIndex] != INF) {
            return dp[depthIndex][colorIndex];
        }
        if (colorIndex == RED) {
            dp[depthIndex][colorIndex] = Math.min(getMinimumCase(dp, cost, depthIndex - 1, GREEN),
                    getMinimumCase(dp, cost, depthIndex - 1, BLUE)
            ) + cost[depthIndex][RED];
        }
        else if (colorIndex == GREEN) {
            dp[depthIndex][colorIndex] = Math.min(getMinimumCase(dp, cost, depthIndex - 1, RED),
                    getMinimumCase(dp, cost, depthIndex - 1, BLUE)
            ) + cost[depthIndex][GREEN];
        }
        else {
            dp[depthIndex][colorIndex] = Math.min(getMinimumCase(dp, cost, depthIndex - 1, RED),
                    getMinimumCase(dp, cost, depthIndex - 1, GREEN)
            ) + cost[depthIndex][BLUE];
        }
        return dp[depthIndex][colorIndex];
    }

    public static void main(String[] args) throws IOException {
//        System.setIn(new FileInputStream("input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st;

        int[][] cost = new int[n][numberOfColor];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int k = 0; k < numberOfColor; k++) {
                cost[i][k] = Integer.parseInt(st.nextToken());
            }
        }

        int[][] dp = new int[n][numberOfColor];
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], INF);
        }

        int startWithRed = getMinimumCase(dp, cost, n - 1, RED);
        int startWithGreen = getMinimumCase(dp, cost, n - 1, GREEN);
        int startWithBlue = getMinimumCase(dp, cost, n - 1, BLUE);

        System.out.println(Math.min(startWithRed, Math.min(startWithGreen, startWithBlue)));
    }
}

/**
 * 기존 코드. 바텀업 방식의 dp 사용
 * dp는 결국 점화식을 뽑아내는게 우선이기에, n일 때의 어떠한 값들이 들어올 수 있는지에 대해 초기에 집중함
 * 결국 문제 조건에서, n번째에는 n-1번째에 들어오지 않은 색만이 들어올 수 있다는 규칙만 지키면, 문제 조건에
 * 위배되는 케이스가 없다고 판단함.
 * 이에 따라 초기에는 단순히 d[n]=d[n-1]+min(n) 정도로 생각을 했지만,
 * 이 점화식이 n+1번째에 어떤 값이 들어오냐에 따른 최소값을 보장할 수 없다는 것을 깨달았다.
 * <p>
 * => 결국 해당 문제에서는 세 가지 상태를 다 봐주어야 한다. n-1번째에서 R, G, B로 오는 경우의 최소값을
 * dp 기법을 통해 저장한 상태로, dp의 점화식 토대를 사용하여, 계속 갱신해주는 로직을 짜야 했다.
 * <p>
 * +) 해당 과정에서 결국 직전의 3가지 상태에 대한 비용값만 필요하기 때문에, 굳이 dp테이블을 만들지 않고,
 * temp[]를 통해, 메모리를 아끼고자 하였다. 이에 따라, 해당 로직은 매 루프마다
 * 기존 비용 temp[]로 복사->최소값 계산->갱신->루프 종료 순으로 이루어진다.
 * <p>
 * +)처음에 점화식을 짠 직 후, 각 케이스별로 보는 것이
 * 브루트포스 알고리즘 성격이 강하다고 생각하여, 아닐 것이라 생각했지만, 브루트포스는 모든 상태를 다 접근해보지만,
 * dp는 n-1번째까지에서 최소값이 되는 케이스만 보고 저장하기 때문에, 살짝 성격이 다르다는 점을 파악했다.
 * (사실 개인적인 생각으로는 브루트포스적 성격이 섞여있는 문제라고 생각이 들었다.)
 */
//public class Main2 {
//    static final int numberOfColor = 3;
//
//    static void getEachMinimumCase(int[] color, int[] temp) {
//        color[0] = Math.min(temp[1], temp[2]);
//        color[1] = Math.min(temp[0], temp[2]);
//        color[2] = Math.min(temp[0], temp[1]);
//    }
//
//    public static void main(String[] args) throws IOException {
/// /        System.setIn(new FileInputStream("input.txt"));
//        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//        int n = Integer.parseInt(br.readLine());
//        StringTokenizer st;
//
//        //R:0 G:1 B:2
//        int[] color = new int[numberOfColor];
//        int[] temp = new int[numberOfColor];
//
//        st = new StringTokenizer(br.readLine());
//        for (int k = 0; k < numberOfColor; k++) {
//            color[k] = Integer.parseInt(st.nextToken());
//        }
//
//        for (int i = 1; i < n; i++) {
//            st = new StringTokenizer(br.readLine());
//
//            System.arraycopy(color, 0, temp, 0, color.length);
//            getEachMinimumCase(color, temp);
//
//            for (int k = 0; k < numberOfColor; k++) {
//                color[k] += Integer.parseInt(st.nextToken());
//            }
//        }
//
//        System.out.println(Arrays.stream(color).min().getAsInt());
//    }
//}

File: 백준/Silver/1149. RGB거리/RGB거리.py
import math
import sys

sys.setrecursionlimit(10 ** 8)


# def get_minimum_cost(cost, n):
#     """
#     코드 1. 바텀업 방식의 dp 구현
#     해당 문제는 주어진 규칙에 따라 n번째의 r,g,b일 떄의 각각의 최소 비용을 구해야 한다.
#     그 이유는 당장 규칙에 따라 최소값을 선택한다하더라도, 그 뒤쪽에 있는 색의 비용이 작다는 보장이 없기 떄문이다.
#     이 조건으로 인해, 그리디 알고리즘을 사용할 수 없으며, 브루트포스 알고리즘으로 모든 경우의 수를 탐색하기에는
#     비용이 보다 많이 필요하기 떄문에, 해당 문제는 n단계에서의 각각의 최소값을 저장하는 dp를 통해 구현한다.
#     dp에 점화식은 다음과 같다.
#     dp[n]=dp[n-1]+cost[r/g/b]
#     n단계에서의 값 3개만 필요하기 떄문에, 별도의 dp테이블을 바텀업 방식에서는 만들지 않고,
#     변수 3개로 롤링 기법을 사용하여 구현할 수 있었따.
# 
#     :param cost: 비용 저장 테이블
#     :param n: 최종 인덱스
#     :return: 최종 최소 비용
#     """
#     r, g, b = cost[0], cost[1], cost[2]
#     for i in range(2, n + 1):
#         k = 3 * (i - 1)
#         r, g, b = min(g, b) + cost[0 + k], min(r, b) + cost[1 + k], min(r, g) + cost[2 + k]
#     return min(r, g, b)
# 
# 
# # sys.stdin = open("input.txt", "r")
# data = sys.stdin.read().split()
# n = int(data[0])
# cost = list(map(int, data[1:]))
# print(get_minimum_cost(cost, n))


# 탑다운
def get_minimum_cost(target_idx, color_idx):
    """
    코드 2. 탑다운 방식의 dp 구현
    바텀업 방식과는 다르게, 탑다운 방식은 메모이제이션을 위해 dp테이블이 필요했다.
    여타 다른 dp 문제와 동일하게 종료 인덱스를 제외한 나머지 인덱스들은 inf로 정의하여,
    재귀의 종료 조건과 메모이제이션을 합쳐 구현할 수 있었다.
    재귀 부분을 정의할 때는, 각 단계에서의 색에 따라 다른 색을 선택해야 하기 때문에,
    color_index와 분기 처리를 통해, 각 단계에서의 규칙을 처리할 수 있었다.
    
    :param target_idx: 각 레벨의 인덱스
    :param color_idx:  각 레벨의 색깔 인덱스(r/g/b: 0/1/2)
    :return: 각 레벨에서의 최소 비용
    """
    if dp[target_idx][color_idx] != math.inf:
        return dp[target_idx][color_idx]
    if color_idx == 0:
        dp[target_idx][color_idx] = min(get_minimum_cost(target_idx - 1, 1)
                                        , get_minimum_cost(target_idx - 1, 2)) + cost_list[target_idx][0]
    if color_idx == 1:
        dp[target_idx][color_idx] = min(get_minimum_cost(target_idx - 1, 0)
                                        , get_minimum_cost(target_idx - 1, 2)) + cost_list[target_idx][1]
    if color_idx == 2:
        dp[target_idx][color_idx] = min(get_minimum_cost(target_idx - 1, 0)
                                        , get_minimum_cost(target_idx - 1, 1)) + cost_list[target_idx][2]
    return dp[target_idx][color_idx]


# sys.stdin = open("input.txt", "r")
data = sys.stdin.read().splitlines()
n = int(data[0])
cost_list = [[] for _ in range(n + 1)]
for i in range(1, n + 1):
    cost_list[i] = list(map(int, data[i].split()))
dp = [[math.inf, math.inf, math.inf] for _ in range(n + 1)]
dp[1][0], dp[1][1], dp[1][2] = cost_list[1][0], cost_list[1][1], cost_list[1][2]
print(min(get_minimum_cost(n, 0), get_minimum_cost(n, 1), get_minimum_cost(n, 2)))
